src/node_port.py

```diff
@@ -151,7 +159,65 @@ def delete_unit_service(self) -> None:
             if e.status.code == 403:
                 self.on_deployed_without_trust()
                 return
-            else:
-                raise
+            raise
+
+    def _node_name(self, unit_name: str) -> str:
+        """Return the node name for this unit's pod ip."""
+        try:
+            pod = self.client.get(
+                Pod,
+                name=unit_name.replace("/", "-"),
+                namespace=self.namespace,
+            )
+        except ApiError as e:
+            if e.status.code == 403:
+                self.on_deployed_without_trust()
+                return
+
+            raise
+
+        return pod.spec.nodeName
+
+    def get_node_ip(self, unit_name: str) -> Optional[str]:
+        """Return node IP for the provided unit."""
+        try:
+            node = self.client.get(
+                Node,
+                name=self._node_name(unit_name),
+                namespace=self.namespace,
+            )
+        except ApiError as e:
+            if e.status.code == 403:
+                logger.error("Could not delete service, application needs `juju trust`")
+                self.on_deployed_without_trust()
+                return
+
+            raise
+        # [
+        #    NodeAddress(address='192.168.0.228', type='InternalIP'),
+        #    NodeAddress(address='example.com', type='Hostname')
+        # ]
+        # Remember that OpenStack, for example, will return an internal hostname, which is not
+        # accessible from the outside. Give preference to ExternalIP, then InternalIP first
+        # Separated, as we want to give preference to ExternalIP, InternalIP and then Hostname
```

Comment:
**praise:** Good comment, explained a question I had.

Reply:
credit where credit is due, you can thank SQL team